### ISSUE TITLE
[BUG] Backup - S3 Timeout

### ISSUE THREAD

**Issue by timmy59100** [longhorn/longhorn#1955]
**Describe the bug**
In the backup UI, I get an error and the backups are not listed.

`error listing backups: error listing backup volumes: Timeout executing: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.0.0/longhorn [backup ls --volume-only s3://PATH/], output , stderr, , error <nil>`

Number of Volumes: 39
Number of Volumes with backup enabled: 32

S3 Bucket Size: 2TB

**Expected behavior**
Backups of Volumes are showed

**Log**
`time="2020-11-05T12:55:15Z" level=warning msg="backup store monitor: failed to list backup volumes in s3://S3PATH/: error listing backup volumes: Timeout executing: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.0.0/longhorn [backup ls --volume-only s3://longhorn-production@ch-dk-2/], output , stderr, , error <nil>"
`

**Environment:**
 - Longhorn version: 1.0.0
 - Kubernetes version: 1.18.3
 - Node OS type and version: Centos 7.7

**Additional context**
On an empty bucket the backups are listed normally.

S3 Provider https://exoscale.com

**Comment by yasker** [comment]
32 volumes is not a big number. And this command only checks the volume itself, so 2TB of data should be out of concern. We need to check how we can reproduce this issue in our lab.

**Comment by jenting** [comment]
From the AWS SDK, the default time is 30 seconds https://github.com/longhorn/backupstore/blob/18aa87a/vendor/github.com/aws/aws-sdk-go/aws/session/cabundle_transport.go#L17
From the issue reported, the longhorn version is `1.0.0` which means that the list operation can't be fulfilled with 30 seconds.

At longhorn release `1.1.0`, we add retry count to `3` https://github.com/longhorn/longhorn-engine/blob/v1.1.0/vendor/github.com/longhorn/backupstore/s3/s3_service.go#L28 which means the list operation tolerance time is 90 seconds.

@timmy59100 Have you tried to upgrade your longhorn cluster to version `1.1.0` to see the error is still happening?

**Comment by timmy59100** [comment]
@jenting Still the same after upgrade to 1.1.0.
It's critical if we can't restore any volumes.

**Comment by jenting** [comment]
@timmy59100 could you please provide the ping latency between your LH cluster to the S3 provider?

**Comment by jenting** [comment]
In my testing
- 3 control plane nodes & 3 worker nodes. Each node with 50G SSD and 4 CPUs and 8G memory
- create 50 volumes and run a recurring backup every 5 mins at all volumes.

If the ping latency from LH cluster to the minio server more than 315ms, click "Backup" on GUI displays **timeout** error.
From the LH manager log, I found that all volumes get at once without pagination.

Therefore, one thing we could improve is when the user clicks "Backup" at GUI, we do not get all volumes at the same time, but with pagination.

**Comment by jenting** [comment]
In my testing
- 3 control plane nodes & 3 worker nodes. Each node with 50G SSD and 4 CPUs and 8G memory
- create 50 volumes and run a recurring backup every 5 mins at all volumes.
If the ping latency from LH cluster to the minio server more than 315ms, click "Backup" on GUI displays **timeout** error.

**Comment by timmy59100** [comment]
@jenting If I just ping the S3 interface it's under 1ms.
Can I measure this s3 specific?

**Comment by jenting** [comment]
@timmy59100 it would be good if you could measure the s3 by aws s3 CLI

1. Install aws s3 in one of the LH manager pod https://docs.aws.amazon.com/cli/latest/userguide/install-linux.html
2. Configure the s3 credential credential
   ```shell
   aws configure
   ```
3. List s3 objects
   ```shell
   time aws s3 ls --endpoint-url <your-s3-endpoint-url> # add --no-verify-ssl if the endpoint is the self-signed certificate
   ```

From my testing, if I add a 1-second delay, the list s3 objects operation increase 3 seconds because 3 HTTP operations performed.

**Comment by timmy59100** [comment]
real    0m0.739s
user    0m0.528s
sys     0m0.077s

**Comment by jenting** [comment]
@timmy59100 could you please log in to one of the longhorn-manager pod and configure the related environment variable to the shell, the environment variable could get from the backup target credential secret (for example, aws-secret).
```shell
kubectl get secret aws-secret -o yaml

export AWS_ACCESS_KEY_ID=<your-aws-access-key-id-in-base64-encoded>
export AWS_SECRET_ACCESS_KEY=<your-aws-secret-access-key-in-base64-encoded>
```

Then, provide the time it consumes to us? thank you.

```shell
time /var/lib/longhorn/engine-binaries/rancher-longhornio-longhorn-engine-v1.1.0/longhorn backup ls --volume-only s3://longhorn-production@ch-dk-2/
```

**Comment by timmy59100** [comment]
![image](https://user-images.githubusercontent.com/7300669/108488627-e529fc80-72a0-11eb-933f-57638276d890.png)

**Comment by dragon2611** [comment]
I had a lot of trouble with this on backblaze as well, it would frequently timeout before bringing up the restore dialogue

**Comment by dragon2611** [comment]
Whilst I really don't recommend this (or at least copy to another bucket or snapshot it first) I did some hackery on my backup repository where I went in and Deleted a load of the .cfg files for the volume.

Obviously that's going to leave you with orphan blocks so you'll probably have to write off the backup repo after but it allowed me to actually restore the latest snapshot.

**Comment by joshimoo** [comment]
@dragon2611we cleanup orphaned blocks when you request a backup delete.
So you can request a backup delete operation from the longhorn ui and we will evaluate all blocks and cleanup orphaned ones.
This will also update the backup volume size to be correct.

**Comment by khushboo-rancher** [comment]
Verified with Longhorn-master - `03/23/2021`

Validation - **Pass**

**With v1.1.0**
Tried listing 50 volumes backups as per https://github.com/longhorn/longhorn/issues/1955#issuecomment-768128322 and it got timed out.
```
[longhorn-manager-rzvxz] time="2021-03-23T17:11:40Z" level=warning msg="HTTP handling error error listing backups: error listing backup volumes: Timeout executing: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://backupbucket@us-east-1/backupstore], output , stderr, , error <nil>" 
[longhorn-manager-rzvxz] time="2021-03-23T17:11:40Z" level=error msg="Error in request: error listing backups: error listing backup volumes: Timeout executing: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.0/longhorn [backup ls --volume-only s3://backupbucket@us-east-1/backupstore], output , stderr, , error <nil>"
```
**With master**
Longhorn listed the backups.

**Comment by jenting** [comment]
To Longhorn maintainers: we probably need to let user configure the timeout value, but the value should not exceed browser timeout (AFAIK, different browser has it's own timeout setting and some browser is not allowed to change it like Chrome).

Or we should implement a sync volume backup controller which periodically sync volume backup (it's a background job so it would not hit the browser timeout). After that, list volume backups from CRs. This also benefit when implement API pagination in Backups page.

**Comment by yasker** [comment]
Btw, you can use @longhorn/maintainer to ping the maintainers.

**Comment by timmy59100** [comment]
@jenting about 30 volumes with 8TB bucket (should not be that big.)
I created a new backup to see if the timeout comes again after some time, like last time.

**Comment by jenting** [comment]
> @jenting about 30 volumes with 8TB bucket (should not be that big.)
> I created a new backup to see if the timeout comes again after some time, like last time.

Thanks. And how many CPUs in the longhorn manager Pods?

**Comment by jenting** [comment]
> @jenting Do we know what's taking the time? It's the listing, or parsing the config files, or anything else? If we can profile it, we might able to fix it from the root, by e.g. using another way to store the cfg files.

I think the problem is single S3 operation needs about 700-800ms in the user's environment _plus_ we have a 2-hashed folder structure.

For example, total 32 volume backups, the worst case is 1 + 32 + 32 = 65 s3 list command queries.
It approximately need 65 * 800ms = 52 secs for list backups.

I think we could choose to remove the 2-hashed folder structure, to make the list of all volume backups in 1 list command query. The problem is the backward-compatible that we should help the user moves the folder structure. So, the folder structure would from
```
0c/5a/pvc-xxx
83/08/pvc-yyy
```
to
```
pvc-xxx
pvc-yyy
```

**Comment by timmy59100** [comment]
@jenting What do you mean by CPUs in Pods?

here is the CPU usage for a longhorn-manager pod:

![image](https://user-images.githubusercontent.com/7300669/118449504-82a39980-b6f3-11eb-8fe5-bc569589ff62.png)

**Comment by jenting** [comment]
I mean the number of CPUs, `cat /proc/cpuinfo`

**Comment by timmy59100** [comment]
8 CPUs.

**Comment by timmy59100** [comment]
@jenting I have timeouts for listing individual backups on a nfs share. Is the mechanism the same as with s3?

error listing backups for volume 'pvc-489cd2ed-6094-42c3-824c-14b0dd9b0c92': error listing backups: Failed to execute: /var/lib/longhorn/engine-binaries/rancher-mirrored-longhornio-longhorn-engine-v1.1.1/longhorn [backup ls --volume pvc-489cd2ed-6094-42c3-824c-14b0dd9b0c92 nfs://PATH], output context deadline exceeded , stderr, time="2021-06-02T08:15:31Z" level=error msg="context deadline exceeded" , error exit status 1

**Comment by luhahn** [comment]
Hi, we also ran into some timeout issues while trying to recover the backups from s3. Is there any possibility to configure this timeout, or is it hard coded?

**Comment by jenting** [comment]
> Hi, we also ran into some timeout issues while trying to recover the backups from s3. Is there any possibility to configure this timeout, or is it hard coded?

It's hard coded :(

**Comment by luhahn** [comment]
Would it be possible to let the user configure this? This is preventing us from recovering backups for large files :(

**Comment by jenting** [comment]
> Would it be possible to let the user configure this? This is preventing us from recovering backups for large files :(

We choose not to let the user configure this is because the browser also has its own timeout setting and some browser like Chrome did not allow the user to change it.

Which page do you encounter the timeout issue on? Click the `Backup` or clicks `Backup` -> `BackupVolume`?

**Comment by luhahn** [comment]
```bash
error listing backups for volume 'nexus-restore': error listing backups: Failed to execute: /var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.1.1/longhorn [backup ls --volume nexus-restore s3://nov000005-km-prod@eu-central-1/production/longhorn/], output context deadline exceeded , stderr, time="2021-07-13T07:45:35Z" level=error msg="context deadline exceeded" , error exit status 1
```
This is the error, and it is shown when I select the backup option in the volume itself:

![grafik](https://user-images.githubusercontent.com/61747797/125412927-7bc8731e-fecd-44e2-bd2a-cc50c95dca02.png)

It is also shown if I go to backups directly und try to select the volume itself in the backups view

**Comment by luhahn** [comment]
I have to admit, I missed that you've already released 1.1.2 :D
Will upgrade later and report back. Thank you for the Info!

We've also cleaned our s3 store of too many *.cfg files and reduced the retention of the backups.
However on 1.1.1 we still run into the timeout. (not always, but like 90% of the time)

**Comment by jenting** [comment]
It'd be good if you could choose the S3 bucket region near your Longhorn cluster, it'll decrease the network latency.
Like you use S3 region `eu-central-1`, it's better to put your Longhorn cluster in Europe.

Also, clean up old backups (back_backup-xxx.cfg) in the S3 bucket manually would be helpful.

**Comment by luhahn** [comment]
Yea, our cluster is located in Europe, so we've already selected the nearest S3 location :)

Shouldn't the retain setting for the backups cleanup old backups?
![grafik](https://user-images.githubusercontent.com/61747797/125416041-bf77c702-41c3-4d15-b81c-ec22ef9e60e9.png)

**Comment by luhahn** [comment]
@jenting Upgrade to 1.1.2 worked. Backups are available all the time now. Thank you! Keep up the good work!

**Comment by innobead** [comment]
Resolved in 1.1.2.
